sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -230,12 +281,271 @@ public ReadFiles withAvroDataModel(GenericData model) {
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        List<BlockMetaData> rowGroups = reader.getRowGroups();
+        long totalSize = 0;
+        for (long i = restriction.getFrom(); i < restriction.getTo(); i++) {
+          totalSize += rowGroups.get((int) i).getTotalByteSize();
+          end += 1;
+          if (totalSize > SPLIT_LIMIT) {
+            start = end;
+            totalSize = 0;
+            out.output(new OffsetRange(start, end));
+          }
+        }
+        if (totalSize != 0) {
+          out.output(new OffsetRange(start, end));
+        }
+      }
+
+      @NewTracker
+      public RestrictionTracker<OffsetRange, Long> newTracker(
+          @Restriction OffsetRange restriction, @Element FileIO.ReadableFile file)
+          throws Exception {
+        return new BlockTracker(
+            restriction, (long) getSize(file, restriction), getRecordCount(file, restriction));
+      }
+
+      @GetRestrictionCoder
+      public OffsetRange.Coder getRestrictionCoder() {
+        return new OffsetRange.Coder();
+      }
+
+      public long getRecordCount(
+          @Element FileIO.ReadableFile file, @Restriction OffsetRange restriction)
+          throws Exception {
+        InputFile inputFile = getInputFile(file);
+        Configuration conf = setConf();
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        if (restriction == null) {
+          return 0;
+        } else {
+          long recordCount = 0;
+          for (long i = restriction.getFrom(); i < restriction.getTo(); i++) {
+            recordCount += reader.getRowGroups().get((int) i).getRowCount();
+          }
+          return recordCount;
+        }
+      }
+
+      public double getSize(@Element FileIO.ReadableFile file, @Restriction OffsetRange restriction)
+          throws Exception {
+        InputFile inputFile = getInputFile(file);
+        Configuration conf = setConf();
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        if (restriction == null) {
+          return 0;
+        } else {
+          double size = 0;
+          for (long i = restriction.getFrom(); i < restriction.getTo(); i++) {
+            size += reader.getRowGroups().get((int) i).getTotalByteSize();
+          }
+          return size;
+        }
+      }
+    }
+
+    private static class BlockTracker extends OffsetRangeTracker {
+      private long totalWork;
+      private long progress;
+      private long recordSize;
+
+      public BlockTracker(OffsetRange range, long totalWork, long recordCount) {
+        super(range);
+        this.recordSize = totalWork / recordCount;
+        this.totalWork = recordSize * recordCount;
```

Comment:
this.totalWork = totalWork

Reply:
I am not sure if there will be a round-off for the record size and the totalWork will be different from the product.